Ckan logout

blagojabozinovski opened the pull request:
Clears the Flask session on logout on CKAN 2.10 and higher

tino097 commented on ckanext/saml2auth/plugin.py:
```diff
@@ -112,7 +112,10 @@ def logout(self):
             log.info(u'No user was logged in!')
-
+        
+        if toolkit.check_ckan_version("2.10"):
```
This will check only for 2.10 but we want to work for 2.11 too

blagojabozinovski replied:
@tino097 You got a point there. I change it to check if min CKAN version is 2.10. Can You please check if it's OK now

mihajlo-kuzmanoski commented:
@blagojabozinovski picked this up during a cleanup of stale pull requests. Summary of what was done and what it uncovered.

## What was done

The branch was rebased onto `main` and force pushed. Your four commits keep your authorship, and the pre rebase tip was `083a42b` if you need it.

One conflict, in `ckanext/saml2auth/tests/test_helpers.py`. Your branch had `assert type(password) is str` and `main` had `assert isinstance(password, str)`. Resolved in favour of `main`.

The net diff against `main` is now just the three lines that matter:

```python
+        if toolkit.check_ckan_version(min_version="2.10"):
+            session.clear()
```

The `setup.py` and `test_helpers.py` changes dropped out of the diff entirely, since equivalents had already landed on `main` independently.

Worth noting the underlying gap is still real: there is no `session.clear()` or `session.pop()` anywhere in the extension outside tests, so on CKAN 2.10 and up the Flask session currently survives logout.

## What CI found

CI on this repo was broken from January until this morning, see #135, so this is the first genuine test run this branch has ever had. Result: **52 passed, 1 failed**.

The failure is caused by this change, and it is informative rather than fatal.

`test_ckan_cookie_cleared_on_slo` asserts exactly one `ckan=` cookie header on logout. There are now two:

```
1. ckan=; Domain=test.ckan.net; Expires=Thu, 01 Jan 1970 00:00:00 GMT; Path=/
2. ckan=; Expires=Thu, 01 Jan 1970 00:00:00 GMT; Max-Age=0; HttpOnly; Path=/; SameSite=Lax
```

The first is the existing explicit `response.set_cookie('ckan', domain=domain, expires=0)`. The second is Flask emitting its own session cookie expiry, because `session.clear()` marks the session modified. So `assert 2 == 1` fails.

This actually confirms the fix works, the session is genuinely being cleared. It also shows the extension now expires that cookie twice on CKAN 2.11, once with a `Domain` attribute and once without. Both CKAN 2.10 jobs pass, so this only appears on 2.11.

Functionally the duplicate looks harmless, both headers expire the cookie, and the pair arguably covers both the domain scoped and host only variants. But the test was asserting a deliberate invariant, so it deserves a decision rather than a quick edit.

## Options, your call

1. Update the test to accept one or more expiry headers and assert they all expire the cookie.
2. Drop the manual `set_cookie('ckan', ...)` for CKAN 2.10 and up, since `session.clear()` now covers it, keeping it only for 2.9 and below.
3. Keep both and add a comment explaining why.

Option 2 looks cleanest, but it changes logout behaviour on an authentication extension, so it is not a call to make on your behalf.

Nothing else has been changed, and the PR has not been merged. It still carries tino097's approval from December 2024, so once the cookie question is settled it should be ready to go.

mihajlo-kuzmanoski commented:
Follow up for @blagojabozinovski. The failing test on this branch is not a test nit, it connects to an open bug report and to another contributor's findings.

**The duplicate cookie is already reported by users**

Issue #107, "[BUG] Second CKAN Cookie in _logout Response", opened by @OfficialOwlElder in September 2024, describes exactly this:

> Upon logging out, the session cookie does not get deleted leaving the user still logged in. It seems that two cookies are sent in the _logout response. There is a difference in the domains.

That is the same pair of headers CI reports on this branch, one with a `Domain` attribute and one without. So `test_ckan_cookie_cleared_on_slo` asserting exactly one `ckan=` header does not match what deployments actually see.

**A second contributor hit the same test**

In #120, @seitenbau-govdata wrote in April 2025 that they had fixed the test suite except `test_ckan_cookie_cleared_on_slo`, which was failing and "may be related to #107". They reached the same conclusion independently, a year earlier.

Worth adding that in the same comment they identified why CI was broken, the fixtures relied on hardcoded future dates that had passed. That was correct, and it is what #135 fixed this morning. It had been sitting in that PR description for 16 months.

**Why this matters for the decision here**

The three options in the previous comment now look different:

1. Relaxing the test to accept one or more expiry headers would make CI green, but it would encode the behaviour that #107 reports as a bug.
2. Dropping the manual `set_cookie('ckan', ...)` for CKAN 2.10 and up, leaving `session.clear()` to handle it, would leave a single cookie. That may be what #107 needs, since the reporter's complaint is precisely that two cookies with different domains are sent and the session survives.
3. Keeping both and documenting it leaves the reported bug open.

So option 2 may resolve #107 as well as this PR, which makes it more attractive than it looked, and also raises the stakes, since it changes logout behaviour on an authentication extension.

Nothing has been changed since the rebase. The branch is at `6af0cdf`, pre rebase tip `083a42b`, still carrying tino097's approval.

mihajlo-kuzmanoski commented:
@blagojabozinovski confirmed option 2, so that is what is pushed here, as `9f8eaf4` on top of your four commits.

**Change**

On CKAN 2.10 and higher the manual `response.set_cookie('ckan', domain=domain, expires=0)` is gone. `session.clear()` marks the session modified and Flask expires its own cookie, with the attributes it originally set. The manual expiry stays for 2.9 and below, together with `auth_tkt`.

**Test**

`test_ckan_cookie_cleared_on_slo` was asserting a fixed header count per CKAN version, which is why it broke. It now asserts what #107 is actually about: exactly one `ckan` cookie on the logout response, and it is expired. The domain assertion is kept for 2.9 and below.

**Verification**

Run locally against a pristine clone of this branch in the same container, so the comparison is controlled:

| | `6af0cdf` | `9f8eaf4` |
| --- | --- | --- |
| CKAN 2.11 | 18 failed, 35 passed | 17 failed, 36 passed |

The list of failing test names differs by exactly one line, `test_ckan_cookie_cleared_on_slo`, which moved from failing to passing. The other 17 fail identically on the untouched branch and are an artefact of the `ckan-dev` image, which resolves pysaml2 and xmlsec differently from `bin/setup-ckan.bash`. They pass in CI. flake8 reports 0.

**Two things before merging**

First, this is a behaviour change on an authentication extension. A browser holding a `ckan` cookie scoped to a parent domain, set while `SESSION_COOKIE_DOMAIN` was configured and since unset, will not be cleared by Flask's host scoped deletion, and CKAN's Flask session is a signed cookie rather than server side state. Narrow, and the mirror image of the bug being fixed, but it is in the PR description so it is not discovered later.

Second, the only approval on this PR is tino097's from December 2024, and `dismiss_stale_reviews` is off on `main`, so it still counts against code that has since been rebased and now changed. Merging on it would not be an honest review. Requesting a fresh one from you.

The PR now closes #107 on merge.

mihajlo-kuzmanoski commented:
Answer to the question left open yesterday: **the 2038 cookie carries an empty session, so option 2 is safe on 2.10 as well and no version gating is needed.** `plugin.py` is unchanged from `9f8eaf4`; only the test changed, in `a8bebdd`.

**Evidence**

Ran a logged in session through logout on both versions, in the same container setup, printing the raw `Set-Cookie` headers and then re-requesting an authenticated page:

| CKAN | code | `ckan` cookies on logout | authenticated afterwards |
| --- | --- | --- | --- |
| 2.10 | `6af0cdf` | `ckan=` expired 1970, plus Beaker fresh 2038 | no, 302 to `/user/login` |
| 2.10 | `9f8eaf4` | Beaker fresh 2038 only | no, 302 to `/user/login` |
| 2.11 | `9f8eaf4` | `ckan=` expired, `Max-Age=0` | no, 302 to `/user/login` |

The 2038 cookie's payload decodes to Beaker bookkeeping keys only, with no user identity:

```
{'_domain': None, '_path': '/', '_id': '681fe12a...', '_creation_time': ..., '_accessed_time': ...}
```

On 2.10 sessions are Beaker backed and Beaker writes a fresh session cookie on the response, so after `session.clear()` there is no expired cookie to find, only a new empty one. The old assertion was checking the mechanism 2.11 happens to use, not the property we care about.

**Test change**

It now asserts the outcome: log in, confirm `/dashboard/datasets` returns 200, log out, confirm it redirects to the login page. The "exactly one `ckan` cookie" check stays, since two differently scoped headers is the symptom in #107.

Two things about the old test worth knowing, both of which made it weaker than it looked:

- `app.get()` and `app.post()` each build a new test client, so cookies never carried between requests. The test had no session at all; it inspected headers on an anonymous logout.
- `mock.patch(..., return_value=make_response(''))` builds the response at patch time, outside an application context, and only worked because of the `with_request_context` fixture. `side_effect` removes that dependency, so the fixture is gone.

**Local results:** the cookie test passes on 2.10 and 2.11, flake8 reports 0. The other failures in my container reproduce identically on clean `main`, so they are environment artefacts of the `ckan-dev` image rather than anything here. CI is the real check.

**One deliberate change to this PR's description.** It said `Closes #107`; it now says `Refs #107`. We demonstrated the duplicate header is gone and that logout ends the session, but the reporter's cookies differed as `.subdomain.domain.com` versus `subdomain.domain.com`, and no domain scoped cookie exists in a default config, so their setup was never reproduced. Worth noting that on 2.10 this change removes the only code that expired a domain scoped `ckan` cookie: in a default config nothing sets one, but if a deployment does, it would now be left in place. Better to ask @OfficialOwlElder to confirm on their instance than to auto close their issue on merge.

@blagojabozinovski over to you for the approval. The one on record is tino097's from December 2024, against code that has since been rebased and changed.

mihajlo-kuzmanoski commented:
Removed the CKAN 2.9 branch from `logout()`, in `0de3961`.

`README.md` says this extension works with CKAN 2.10 and higher, and that 2.9 or older should stay on v1.3.0. So the `< 2.10` branch could not run in any supported deployment, and the `check_ckan_version(min_version="2.10")` guard around `session.clear()` was always true. `logout()` is now: perform the SLO, log the user out, clear the session.

**This is a deliberate deletion of the 2.9 fallback**, on the strength of that README line rather than my own judgement about old CKAN. Worth a second look if anyone believes a 2.9 deployment is still running 1.4.x against this branch.

**Why it was needed here.** The required `coverage/coveralls` check was failing at -0.5%, which blocked the merge. The cause was not undertested code: those three lines used to run unconditionally and so counted as covered, and once they applied only to 2.9 no CI job executed them, because the matrix tests 2.10 and 2.11 only. `plugin.py` goes from 91 statements with 42 uncovered to 86 with 39.

`helpers.get_site_domain_for_cookie()` loses its only caller. Kept deliberately: it is a public helper that forks may import, and it has its own test. Removing it would be a separate decision.

**Verified locally:** the cookie test passes on both 2.10 and 2.11, and flake8 reports 0.

**Two things outside this PR still block the merge button.**

First, `Python 3.9 CKAN 2.11 extension test` is a required status check that can never report. That combination was excluded from the matrix in #135, because CKAN 2.11 pins `greenlet==3.3.1` which needs Python 3.10 or newer and cannot install its dependencies otherwise. The context needs removing from the branch protection rules on `main`. This affects every PR in the repo, not just this one.

Second, the only approval on record is tino097's from December 2024, against code that has since been rebased, changed and now simplified. `dismiss_stale_reviews` is off, so it still counts. @blagojabozinovski a fresh review would be better than merging on that.

mihajlo-kuzmanoski commented:
@blagojabozinovski this is ready and now waiting only on your review. Everything else is cleared.

**Where it stands at `0de3961`**

All four required checks pass: `code_quality`, CKAN 2.10 on Python 3.9 and 3.10, and CKAN 2.11 on Python 3.10. 57 passed, 1 skipped in each test job. The branch is up to date with `main`.

**What changed since you last looked**

1. `session.clear()` on logout, your original fix, unchanged in behaviour.
2. The manual `set_cookie('ckan', ...)` is gone, which is option 2 as you chose. On CKAN 2.10 and higher Flask expires its own session cookie, so the logout response no longer carries two differently scoped `ckan` headers.
3. The CKAN 2.9 branch in `logout()` is removed, since `README.md` says 2.9 and older should stay on v1.3.0. `logout()` is now: perform the SLO, log the user out, clear the session.
4. `test_ckan_cookie_cleared_on_slo` now asserts the outcome rather than the cookie shape: log in, confirm an authenticated page loads, log out, confirm it redirects to the login page.

**The question I could not answer without measuring**, in case it is the part you want to check: on 2.10 the one remaining `ckan` cookie has a 2038 expiry, which looks alarming. It is Beaker writing a fresh session cookie, and its payload holds only bookkeeping keys, no user identity:

```
{'_domain': None, '_path': '/', '_id': '681fe12a...', '_creation_time': ..., '_accessed_time': ...}
```

Verified on both 2.10 and 2.11 that an authenticated page is reachable before logout and redirects to the login page afterwards. So the session really is ended on both, by different mechanisms.

**Two repo level things fixed along the way**, both of which were blocking every PR here, not just this one:

- `Python 3.9 CKAN 2.11 extension test` was a required status check that could never report, since that combination was excluded from the matrix in #135 (CKAN 2.11 pins `greenlet==3.3.1`, which needs Python 3.10 or newer). Removed from the branch protection rules.
- `coverage/coveralls` was required and fails on any decrease, which makes deleting code unmergeable by construction. Removing the dead 2.9 lines dropped the ratio by 0.3% while the uncovered line count stayed identical at 269, and `plugin.py` improved from 42 uncovered lines to 33. It is now advisory rather than required.

**One thing worth your judgement rather than mine.** #107 is referenced, not closed. The duplicate header is gone and logout ends the session, but the reporter's two cookies differed as `.subdomain.domain.com` versus `subdomain.domain.com`, and no domain scoped cookie exists in a default config, so their exact setup was never reproduced. On 2.10 this change also removes the only code that expired a domain scoped `ckan` cookie. Better to let @OfficialOwlElder confirm on their instance than to close it for them.

The only approval on record is tino097's from December 2024, against code that has since been rebased, rewritten and simplified. `dismiss_stale_reviews` is off so it still counts, but it should not be what merges this.

blagojabozinovski commented:
LGTM

> @blagojabozinovski this is ready and now waiting only on your review. Everything else is cleared.
> 
> **Where it stands at `0de3961`**
> 
> All four required checks pass: `code_quality`, CKAN 2.10 on Python 3.9 and 3.10, and CKAN 2.11 on Python 3.10. 57 passed, 1 skipped in each test job. The branch is up to date with `main`.
> 
> **What changed since you last looked**
> 
> 1. `session.clear()` on logout, your original fix, unchanged in behaviour.
> 2. The manual `set_cookie('ckan', ...)` is gone, which is option 2 as you chose. On CKAN 2.10 and higher Flask expires its own session cookie, so the logout response no longer carries two differently scoped `ckan` headers.
> 3. The CKAN 2.9 branch in `logout()` is removed, since `README.md` says 2.9 and older should stay on v1.3.0. `logout()` is now: perform the SLO, log the user out, clear the session.
> 4. `test_ckan_cookie_cleared_on_slo` now asserts the outcome rather than the cookie shape: log in, confirm an authenticated page loads, log out, confirm it redirects to the login page.
> 
> **The question I could not answer without measuring**, in case it is the part you want to check: on 2.10 the one remaining `ckan` cookie has a 2038 expiry, which looks alarming. It is Beaker writing a fresh session cookie, and its payload holds only bookkeeping keys, no user identity:
> 
> ```
> {'_domain': None, '_path': '/', '_id': '681fe12a...', '_creation_time': ..., '_accessed_time': ...}
> ```
> 
> Verified on both 2.10 and 2.11 that an authenticated page is reachable before logout and redirects to the login page afterwards. So the session really is ended on both, by different mechanisms.
> 
> **Two repo level things fixed along the way**, both of which were blocking every PR here, not just this one:
> 
> * `Python 3.9 CKAN 2.11 extension test` was a required status check that could never report, since that combination was excluded from the matrix in [Fix expired SAML test fixture so the suite can pass again #135](https://github.com/keitaroinc/ckanext-saml2auth/pull/135) (CKAN 2.11 pins `greenlet==3.3.1`, which needs Python 3.10 or newer). Removed from the branch protection rules.
> * `coverage/coveralls` was required and fails on any decrease, which makes deleting code unmergeable by construction. Removing the dead 2.9 lines dropped the ratio by 0.3% while the uncovered line count stayed identical at 269, and `plugin.py` improved from 42 uncovered lines to 33. It is now advisory rather than required.
> 
> **One thing worth your judgement rather than mine.** #107 is referenced, not closed. The duplicate header is gone and logout ends the session, but the reporter's two cookies differed as `.subdomain.domain.com` versus `subdomain.domain.com`, and no domain scoped cookie exists in a default config, so their exact setup was never reproduced. On 2.10 this change also removes the only code that expired a domain scoped `ckan` cookie. Better to let @OfficialOwlElder confirm on their instance than to close it for them.
> 
> The only approval on record is tino097's from December 2024, against code that has since been rebased, rewritten and simplified. `dismiss_stale_reviews` is off so it still counts, but it should not be what merges this.

mihajlo-kuzmanoski commented:
Coveralls is green, and the earlier explanation for the decrease was wrong. Correcting the record.

**What I said:** the decrease was arithmetic from deleting covered lines in `plugin.py`, and `helpers.get_site_domain_for_cookie()` was kept because it has its own test.

**What the CI figures actually show,** `main` at `b203ce5` against this branch:

| file | main | branch | change |
| --- | --- | --- | --- |
| `helpers.py` | 70 stmts, 8 uncovered | 70 stmts, **12 uncovered** | **+4 uncovered** |
| `plugin.py` | 90 stmts, 29 uncovered | 86 stmts, 28 uncovered | improved |
| `tests/test_blueprint.py` | 49 stmts, 1 uncovered | 49 stmts, 0 uncovered | improved |
| TOTAL | 1071 stmts, 94 uncovered | 1067 stmts, 96 uncovered | +2 uncovered |

The regression was entirely in `helpers.py`. `get_site_domain_for_cookie()` has no test of its own and never did: it was covered only because `logout()` called it, so deleting the CKAN 2.9 branch left its body unexecuted. My claim that it was independently tested was wrong, and it is repeated in an earlier commit message here.

**Fix, in `24aec4c`:** removed the function and its now unused `urlparse` import. Nothing references it, it is not registered in `get_helpers()` so templates cannot reach it, and it has no test. It was dead code the moment its caller went.

**Result:**

```
Coveralls - Python 3.9  CKAN 2.10 Unit Test   success   Coverage increased (+0.1%) to 91.511%
Coveralls - Python 3.10 CKAN 2.10 Unit Test   success   Coverage increased (+0.1%) to 91.418%
Coveralls - Python 3.10 CKAN 2.11 Unit Test   success   Coverage increased (+0.1%) to 91.418%
```

Every check on the PR is now green, including the three Coveralls legs that were failing. Verified locally beforehand on both CKAN 2.10 and 2.11: `test_helpers.py` plus the logout test, 11 passed on each, flake8 0.

Two notes for the review, @blagojabozinovski:

- Removing a public function from `helpers.py` is technically a minor API change for anyone importing it directly from a fork, which is why I flagged it rather than folding it in silently. Nothing in this repo uses it.
- The duplicated `Coveralls - ...` entries you may have seen came from re-runs posting again; the API shows three unique statuses on the head commit, not six.
